lib/pure/os.nim

```diff
@@ -3217,6 +3184,44 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
+proc sameFileContent*(path1, path2: string; checkSize = false; bufferSize = 8192.Positive): bool {.
+    rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns true if both pathname arguments refer to files with identical
+  ## binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ##
+  ## .. code-block:: nim
+  ##   echo sameFileContent("file0.txt", "file1.txt", checkSize = true, bufferSize = 4096)
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not(open(a, path1)) or not(open(b, path2)): mustRead = false
+    if mustRead and checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
+    if mustRead:
+      while true:
+        var readA = readBuffer(a, bufA, bufferSize)
+        var readB = readBuffer(b, bufB, bufferSize)
+        if readA != readB: break
+        if readA == 0:
```

Comment:
seems incorrect; 0 could indicate error instead of end of file:
man fread:
> If an error occurs, or the end-of-file is reached, the return value is a short object count (or zero).

instead how about:
```nim
        if readA > 0:
          result = equalMem(bufA, bufB, readA)
          if not result: break
        if readA != bufferSize:
          let enda = io.endOfFile(a)
          let endb = io.endOfFile(b)
          if enda != endb: break
          if enda:
            result = true
            break
```